src/qtt/measurements/scans.py

```diff
@@ -870,22 +870,41 @@ def _convert_scanjob_vec(self, station, steplength=None, sweeplength=None, stepv
         """
 
         def _update_step_size_from_length(data, length):
-            """ From desired number of data points determine the required step size """
+            """ From desired number of data points determine the required step size.
+            length > 1: to get 'inclusive' data with length = 1, start and end values must be the same and step value
+            must be 0 which are invalid values.
+            """
             if length is not None:
                 if length == 1:
-                    data['step'] = 0
+                    raise ValueError('length must be > 1')
                 else:
                     if 'range' in sweepdata:
                         data['step'] = data['range'] / (length-1)
                     else:
                         data['step'] = (data['end'] - data['start']) / (length-1)
 
-        def _calculate_sweepvalues(param, sweep_data) -> SweepFixedValues:
-            """ From parameter and sweep data create a SweepFixedValues object """
-            sweep_values = SweepFixedValues(param,
-                                            start=sweep_data['start'],
-                                            stop=sweep_data['end'],
-                                            step=sweep_data['step'])
+        def _calculate_sweepvalues(param, sweep_data, inclusive_end=False) -> SweepFixedValues:
+            """ From parameter and sweep data create a SweepFixedValues object. """
+            if sweep_data['step'] == 0.:
+                raise ValueError('step value may not be 0')
+            steps = abs((sweep_data['end'] - sweep_data['start']) / sweep_data['step'])
+            tolerance = 1e-10
+            steps_lo = int(np.floor(steps + tolerance))
+            steps_hi = int(np.ceil(steps - tolerance))
+
+            if inclusive_end:
+                if steps_lo != steps_hi:
+                    sweep_data['end'] = sweep_data['start'] + steps_lo * sweep_data['step']
```

Comment:
This modifies the `sweep_data`. That is a bit shady, unless already a deepcopy was made before.

Reply:
The dictionary sweep_data is passed during construction. So, any changes could have external effects.
sweep_data is also changed in many other places, e.g. lines 914, 915, 919,920, 924, 975, ...

Reply:
You are right, this mutable parameter is adjusted in a lot of places. What is different is that this is an extra adjustment so in the end this new code has some side effect that was not in before.
I've thought about what I was trying to fix for this part of the original change but I think I am doing more bad than good. So what I will do is put back the old code for this 'inclusive end' part and only do the checks because these checks will lead to more understandable error messages.